iree/vm/list.h

```diff
@@ -0,0 +1,127 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IREE_VM_LIST_H_
+#define IREE_VM_LIST_H_
+
+#include <stdint.h>
+
+#include "iree/base/api.h"
+#include "iree/vm/ref.h"
+#include "iree/vm/type_def.h"
+#include "iree/vm/value.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+// A growable list that can hold primitive value types or ref objects or a mix.
+// This maps to the `!vm.list<...>` type in the VM IR and is designed to enable
+// flexible interop between hosting applications using the VM C API to invoke IR
+// and custom modules that need to pass arbitrary list-like data across the VM
+// ABI. It is not designed for efficiency: if you are performing large amounts
+// of work on the list type you should instead be representing that using the
+// HAL types so that you can get acceleration.
```

Comment:
A sampling profiler/inspection would be the only way - it's like using a CHECK'd std::vector instead of std::array: each action just takes a tiny bit longer.